File: topicos/dynamic-programming/group-1/geometric_series/geometric_series.ts
export function geometric_series(n: bigint, a1: bigint, q: bigint) {
    if (q == 1n) {
        return n * a1;
    }
    
    return a1 * (binary_exponation(q, n) - 1 ) / (q-1)
}


function binary_exponation(base: bigint, exp: bigint): bigint {
    let result: bigint = 1n;

    while (exp > 0n) {
        if ((exp&1n) === 1n) {
            result *= base;
        }

        base *= base;
        exp >>= 1n;
    }

    return result;
}

File: topicos/dynamic-programming/group-1/geometric_series/geometric_series.go
package main

import "fmt"

func binaryExponation(base: int64, exp: int64): int64 {
	var result: int64 = 1

	for exp > 0 {
		if exp&1 == 1 {
			result *= base
		}

		base *= base
		result >>= 1
	}

	return result
}

func geometricSeries(n: int64, a1: int64, q: int64): int64 {
	if q == 1 {
		return a1 * n
	}

	return a1 * (binaryExponation(q, n) - 1) / q -1
}